### DISCUSSION TITLE
Program in MATLAB and connect via Bluetooth

### DISCUSSION THREAD

**Discussion by erodriguezseda** [iRobotEducation/create3_docs#206] [Questions] [answered]
Hello,
Is there a way of sending command to iRobot create 3 via Bluetooth without using Python Web Playground or ROS? In particular, I would like to use MATLAB to send commands as I have used it with previous versions of the iRobot. One issue I find is that the iRobot is not discoverable by my laptop except when I use the iRobot coding app or the web application. 
Thanks!

**Comment by shamlian** [comment, marked as answer]
You are welcome to write your own library to communicate with the robot, as the [protocol is fully documented here](https://github.com/iRobotEducation/root-robot-ble-protocol#). Alternatively, you are welcome to use the iRobot downloadable Python SDK, which is available [here](https://github.com/iRobotEducation/irobot-edu-python-sdk/tree/main).

Please note also that the Create 3 robot uses Bluetooth Low Energy, and not Bluetooth Classic, so it doesn't need to "pair" like a Bluetooth headset or similar device.

[I don't know anything about programming BLE devices in MATLAB, but it looks possible.](https://www.mathworks.com/help/matlab/import_export/bluetooth-low-energy-communication-overview.html)

**Comment by erodriguezseda** [comment]
Thanks a lot! I am now able to find and connect to the robot from MATLAB. I am following this [documentation](https://github.com/iRobotEducation/root-robot-ble-protocol#) you gave me and so far I am able to read data from the robot (when an event takes place) using the read command in MATLAB. I am having issues sending a command (to request battery level or send a motor speed). I wonder, does the robot need to be set up in a particular mode to receive message (despite it can send)? Do I need to write the crc-8 bit in the message (or can disable it)? Thanks.

PS: I sent a request for the downloadable Python SDK, looking forward to it.

**Reply by shamlian** [reply]
That's exciting! Yes, you need to write the CRC for the robot to accept the message. I believe the algorithm is a standard one, but an implementation can be found under "Checksum" [here](https://github.com/iRobotEducation/root-robot-ble-protocol#packet-components). I'm guessing MATLAB has an implementation (because they have an implementation for everything, right? :) ). The robot doesn't have to be in any mode to accept packets -- if you're receiving them on sensor events, it should also accept commands.

**Reply by shamlian** [reply]
You can also get some sample packets to validate your algorithm from [this old Python Root library](https://github.com/zlite/PyRoot/blob/master/drive-root.py#L67,L80) written by a third party.

**Reply by erodriguezseda** [reply]
THANK YOU! I tried some of the "sample packets" and was able to move/stop the robot. I just need now to figure out which CRC8 algorithm I used (which I think I know) and how to implement it in MATLAB. Thanks for your help.

**Reply by shamlian** [reply]
You're very welcome! I'm excited to see your MATLAB library, if you choose to publish it! (Also -- the library you write will be compatible with Root robots as well, which could be pretty interesting.)

**Reply by iMachover** [reply]
How did you use this link https://github.com/iRobotEducation/root-robot-ble-protocol# to figure it out? it seems gibberish to me, I am going to try it with [this documentation](https://www.mathworks.com/help/matlab/import_export/bluetooth-low-energy-communication-overview.html) though tomorrow. it looks like it shows how to do it in matlab as opposed to the other link

**Reply by shamlian** [reply]
I'm unsure if this is helpful, but I'll try: the iRobot Education protocol document describes how to use Bluetooth Low Energy concepts to control the robot. The Mathworks document describes how to use Bluetooth Low Energy with Matlab. Once you understand how BLE works, re-read the Overview and Profile sections of the iRobot document to see what characteristics and services are required to communicate with the robot; then, you exchange packets with the robot as described in the lower sections to make it do stuff. Maybe @erodriguezseda would be willing to share some of their code to help you out? You could also look at the source of our [Python SDK](https://github.com/iRobotEducation/irobot-edu-python-sdk) for implementation hints.

**Reply by iMachover** [reply]
Hey thanks, I was able to connect via BLE ex: b = ble("iRobot"); but the services and characteristics are very limited.  I stored cahracter customName and customUID to = c. None of the other characteristics  can initialize, I kept getting a matrix error. I tried reading data from a bumper, well it worked but it said there was no data  even though it sent that data when the bumper was pressed, I got 0x00. I'll try to figure out the protocol but it seems unlikely, I'm brand new to Matlab and robotics. All I need it to do is read data, and move and eventually control it with an xbox controller. Seems light years away at this point. I appreciate the reply though.

**Comment by erodriguezseda** [comment]
Hello, 
Yes, I can share some functions I wrote in MATLAB. They should allow you to connect to the robot, send velocity commands, stop, and get odometry information. I created a [repository](https://github.com/erodriguezseda/iRobot-Create-3-/). Hope they help you.

**Reply by iMachover** [reply]
Thanks, we actually decided to go with ROS2 for now, but may in the future try out BLE. I will definitely use this when it gets to that point. I appreciate the help if I have any other questions I will definitely be back.
